File: example/module/mnist_mlp.py
# pylint: skip-file
import mxnet as mx
import numpy as np
import logging

data = mx.symbol.Variable('data')
fc1 = mx.symbol.FullyConnected(data, name='fc1', num_hidden=128)
act1 = mx.symbol.Activation(fc1, name='relu1', act_type="relu")
fc2 = mx.symbol.FullyConnected(act1, name = 'fc2', num_hidden = 64)
act2 = mx.symbol.Activation(fc2, name='relu2', act_type="relu")
fc3 = mx.symbol.FullyConnected(act2, name='fc3', num_hidden=10)
softmax = mx.symbol.SoftmaxOutput(fc3, name = 'softmax')

n_epoch = 2
batch_size = 100
train_dataiter = mx.io.MNISTIter(
        image="../image-classification/mnist/train-images-idx3-ubyte",
        label="../image-classification/mnist/train-labels-idx1-ubyte",
        data_shape=(784,),
        batch_size=batch_size, shuffle=True, flat=True, silent=False, seed=10)
val_dataiter = mx.io.MNISTIter(
        image="../image-classification/mnist/t10k-images-idx3-ubyte",
        label="../image-classification/mnist/t10k-labels-idx1-ubyte",
        data_shape=(784,),
        batch_size=batch_size, shuffle=True, flat=True, silent=False)

################################################################################
# Intermediate-level API
################################################################################
mod = mx.mod.Module(softmax)
mod.bind(data_shapes=train_dataiter.provide_data, label_shapes=train_dataiter.provide_label)
mod.init_params()

mod.init_optimizer(optimizer_params={'learning_rate':0.01, 'momentum': 0.9})
metric = mx.metric.create('acc')

for i_epoch in range(n_epoch):
    for i_iter, batch in enumerate(train_dataiter):
        mod.forward(batch)
        mod.update_metric(metric, batch.label)

        mod.backward()
        mod.update()

    for name, val in metric.get_name_value():
        print('epoch %03d: %s=%f' % (i_epoch, name, val))
    metric.reset()
    train_dataiter.reset()


################################################################################
# High-level API
################################################################################
logging.basicConfig(level=logging.DEBUG)
train_dataiter.reset()
mod = mx.mod.Module(softmax)
mod.fit(train_dataiter, eval_data=val_dataiter,
        optimizer_params={'learning_rate':0.01, 'momentum': 0.9}, num_epoch=n_epoch)

# prediction iterator API
for preds, i_batch, batch in mod.iter_predict(val_dataiter):
    pred_label = preds[0].asnumpy().argmax(axis=1)
    label = batch.label[0].asnumpy().astype('int32')
    if i_batch % 20 == 0:
        print('batch %03d acc: %.3f' % (i_batch, (label == pred_label).sum() / float(len(pred_label))))

# a dummy call just to test if the API works for merge_batches=True
preds = mod.predict(val_dataiter)

# perform prediction and calculate accuracy manually
preds = mod.predict(val_dataiter, merge_batches=False)
val_dataiter.reset()
acc_sum = 0.0; acc_cnt = 0
for i, batch in enumerate(val_dataiter):
    pred_label = preds[i][0].asnumpy().argmax(axis=1)
    label = batch.label[0].asnumpy().astype('int32')
    acc_sum += (label == pred_label).sum()
    acc_cnt += len(pred_label)
print('validation Accuracy: %.3f' % (acc_sum / acc_cnt))

# evaluate on validation set with a evaluation metric
mod.score(val_dataiter, metric)
for name, val in metric.get_name_value():
    print('%s=%f' % (name, val))

File: scala-package/examples/src/main/scala/ml/dmlc/mxnetexamples/module/MnistMlp.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ml.dmlc.mxnetexamples.module

import ml.dmlc.mxnet._
import ml.dmlc.mxnet.module.{FitParams, Module}
import ml.dmlc.mxnet.DataDesc._
import ml.dmlc.mxnet.optimizer.SGD
import org.kohsuke.args4j.{Option, CmdLineParser}
import org.slf4j.LoggerFactory

import scala.collection.JavaConverters._

object MnistMlp {
  private val logger = LoggerFactory.getLogger(classOf[MnistMlp])

  def getSymbol: Symbol = {
    val data = Symbol.Variable("data")
    val fc1 = Symbol.FullyConnected(name = "fc1")(data)(Map("num_hidden" -> 128))
    val act1 = Symbol.Activation(name = "relu1")(fc1)(Map("act_type" -> "relu"))
    val fc2 = Symbol.FullyConnected(name = "fc2")(act1)(Map("num_hidden" -> 64))
    val act2 = Symbol.Activation(name = "relu2")(fc2)(Map("act_type" -> "relu"))
    val fc3 = Symbol.FullyConnected(name = "fc3")(act2)(Map("num_hidden" -> 10))
    val softmax = Symbol.SoftmaxOutput(name = "softmax")(fc3)()
    softmax
  }

  def runIntermediateLevelApi(train: DataIter, eval: DataIter,
      cmdLine: MnistMlp, loadModelEpoch: Int = -1): Unit = {
    // Intermediate-level API
    val mod = if (loadModelEpoch == -1) {
      new Module(getSymbol)
    } else {
      logger.info("Load checkpoint from epoch {}", loadModelEpoch)
      Module.loadCheckpoint("model/mnist_mlp", loadModelEpoch, loadOptimizerStates = true)
    }
    mod.bind(dataShapes = train.provideData, labelShapes = Some(train.provideLabel))
    mod.initParams()
    mod.initOptimizer(optimizer = new SGD(learningRate = 0.01f, momentum = 0.9f))

    val metric = new Accuracy()

    for (epoch <- 0 until cmdLine.numEpoch) {
      while (train.hasNext) {
        val batch = train.next()
        mod.forward(batch)
        mod.updateMetric(metric, batch.label)
        mod.backward()
        mod.update()
      }

      mod.saveCheckpoint("model/mnist_mlp", epoch, saveOptStates = true)

      val (name, value) = metric.get
      name.zip(value).foreach { case (n, v) =>
        logger.info(s"epoch $epoch $n=$v")
      }
      metric.reset()
      train.reset()
    }
  }

  def runHighLevelApi(train: DataIter, test: DataIter, cmdLine: MnistMlp): Unit = {
    // High-level API
    train.reset()
    val mod = new Module(getSymbol)
    mod.fit(train, evalData = scala.Option(test), numEpoch = cmdLine.numEpoch)

    // prediction iterator API
    var iBatch = 0
    test.reset()
    while (test.hasNext) {
      val batch = test.next()
      val preds = mod.predict(batch)
      val predLabel: Array[Int] = NDArray.argmax_channel(preds(0)).toArray.map(_.toInt)
      val label = batch.label(0).toArray.map(_.toInt)
      val acc = predLabel.zip(label).map { case (py, y) =>
        if (py == y) 1 else 0
      }.sum / predLabel.length.toFloat
      if (iBatch % 20 == 0) {
        logger.info(s"Batch $iBatch acc: $acc")
      }
      iBatch += 1
    }

    // a dummy call just to test if the API works
    mod.predict(test)

    // perform prediction and calculate accuracy manually
    val preds = mod.predictEveryBatch(test)
    test.reset()
    var accSum = 0.0f
    var accCnt = 0
    var i = 0
    while (test.hasNext) {
      val batch = test.next()
      val predLabel: Array[Int] = NDArray.argmax_channel(preds(i)(0)).toArray.map(_.toInt)
      val label = batch.label(0).toArray.map(_.toInt)
      accSum += (predLabel zip label).map { case (py, y) =>
        if (py == y) 1 else 0
      }.sum
      accCnt += predLabel.length
      i += 1
    }
    logger.info(s"Validation Accuracy: {}", accSum / accCnt.toFloat)

    // evaluate on validation set with a evaluation metric
    val (name, value) = mod.score(test, new Accuracy).get
    logger.info("Scored {} = {}", name(0), value(0))
  }

  def main(args: Array[String]): Unit = {
    val inst = new MnistMlp
    val parser = new CmdLineParser(inst)
    try {
      parser.parseArgument(args.toList.asJava)

      val train = IO.MNISTIter(Map(
        "image" -> (inst.dataDir + "train-images-idx3-ubyte"),
        "label" -> (inst.dataDir + "train-labels-idx1-ubyte"),
        "label_name" -> "softmax_label",
        "input_shape" -> "(784,)",
        "batch_size" -> inst.batchSize.toString,
        "shuffle" -> "True",
        "flat" -> "True", "silent" -> "False", "seed" -> "10"))
      val eval = IO.MNISTIter(Map(
        "image" -> (inst.dataDir + "t10k-images-idx3-ubyte"),
        "label" -> (inst.dataDir + "t10k-labels-idx1-ubyte"),
        "label_name" -> "softmax_label",
        "input_shape" -> "(784,)",
        "batch_size" -> inst.batchSize.toString,
        "flat" -> "True", "silent" -> "False"))

      logger.info("Run intermediate level api from beginning.")
      runIntermediateLevelApi(train, eval, inst)
      logger.info("Run intermediate level api, start with last trained epoch.")
      runIntermediateLevelApi(train, eval, inst, loadModelEpoch = inst.numEpoch - 1)
      logger.info("Run high level api")
      runHighLevelApi(train, eval, inst)
    } catch {
      case ex: Exception =>
        logger.error(ex.getMessage, ex)
        parser.printUsage(System.err)
        sys.exit(1)
    }
  }
}

class MnistMlp {
  @Option(name = "--data-dir", usage = "the input data directory")
  private val dataDir: String = "mnist/"
  @Option(name = "--batch-size", usage = "the batch size for data iterator")
  private val batchSize: Int = 2
  @Option(name = "--num-epoch", usage = "number of training epoches")
  private val numEpoch: Int = 10
}
